internal/pkg/provider/provision.go

```diff
@@ -0,0 +1,276 @@
+	"github.com/siderolabs/omni/client/pkg/constants"
+	"github.com/siderolabs/omni/client/pkg/infra/provision"
+	"github.com/siderolabs/omni/client/pkg/omni/resources/infra"
+	"go.uber.org/zap"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	kvv1 "kubevirt.io/api/core/v1"
+	cdiv1 "kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/siderolabs/omni-infra-provider-kubevirt/internal/pkg/provider/resources"
+)
+
+// Provisioner implements Talos emulator infra provider.
+type Provisioner struct {
+	k8sClient client.Client
+	namespace string
+}
+
+// NewProvisioner creates a new provisioner.
+func NewProvisioner(k8sClient client.Client, namespace string) *Provisioner {
+	return &Provisioner{
+		k8sClient: k8sClient,
+		namespace: namespace,
+	}
+}
+
+// ProvisionSteps implements infra.Provisioner.
+//
+//nolint:gocognit
+func (p *Provisioner) ProvisionSteps() []provision.Step[*resources.Machine] {
+	return []provision.Step[*resources.Machine]{
+		provision.NewStep("ensureVolume", func(ctx context.Context, logger *zap.Logger, pctx provision.Context[*resources.Machine]) error {
+			schematic, err := pctx.GenerateSchematicID(ctx, logger, provision.WithExtraKernelArgs("console=ttyS0,38400n8"))
+			if err != nil {
+				return err
+			}
+
+			pctx.State.TypedSpec().Value.Schematic = schematic
+			pctx.State.TypedSpec().Value.TalosVersion = pctx.GetTalosVersion()
+
+			url, err := url.Parse(constants.ImageFactoryBaseURL)
+			if err != nil {
+				return err
+			}
+
+			var data Data
+
+			err = pctx.UnmarshalProviderData(&data)
+			if err != nil {
+				return err
+			}
+
+			url = url.JoinPath("image", schematic, pctx.GetTalosVersion(), fmt.Sprintf("metal-%s.raw", data.Architecture))
+
+			hash := sha256.New()
+
+			if _, err = hash.Write([]byte(url.String())); err != nil {
+				return err
+			}
+
+			volumeID := hex.EncodeToString(hash.Sum(nil))
+
+			pctx.State.TypedSpec().Value.VolumeId = volumeID
+
+			volume := cdiv1.DataVolume{
+				Spec: cdiv1.DataVolumeSpec{
+					Source: &cdiv1.DataVolumeSource{
+						HTTP: &cdiv1.DataVolumeSourceHTTP{
+							URL: url.String(),
+						},
+					},
+					PVC: &v1.PersistentVolumeClaimSpec{
+						VolumeMode: pointer.To(v1.PersistentVolumeBlock),
+						AccessModes: []v1.PersistentVolumeAccessMode{
+							v1.ReadWriteOnce,
+						},
+						Resources: v1.VolumeResourceRequirements{
+							Requests: v1.ResourceList{
+								v1.ResourceStorage: resource.MustParse("1.5Gi"),
+							},
+						},
+					},
+				},
+			}
+
+			vol := &cdiv1.DataVolume{}
+
+			err = p.k8sClient.Get(ctx, client.ObjectKey{
+				Namespace: p.namespace,
+				Name:      volumeID,
+			}, vol)
+			if err != nil && !errors.IsNotFound(err) {
+				return err
+			}
+
+			if vol.Status.Phase == cdiv1.Succeeded {
+				return nil
```

Comment:
and is Status is !Succeeded, we proceed with Create? this doesn't look correct, and should only be checked if err == nil I guess